### PR TITLE
Fix race problem in configuration tests

### LINKED ISSUE #86532: Microsoft.Extensions.Configuration.Binder.Tests.ConfigurationBinderTests.BindCanReadStaticProperty failing intermittently
## Build Information
Build: https://dev.azure.com/dnceng-public/cbb18261-c48f-4abb-8651-8cdcb5474649/_build/results?buildId=280353
Build error leg or test failing: Microsoft.Extensions.Configuration.Binder.Tests.ConfigurationBinderTests.BindCanReadStaticProperty
Pull request: https://github.com/dotnet/runtime/pull/86519

## Error Message
Fill the error message using [known issues guidance](https://github.com/dotnet/arcade/blob/main/Documentation/Projects/Build%20Analysis/KnownIssues.md#how-to-fill-out-a-known-issue-error-section).

```json
{
  "ErrorMessage": "Expected: other stuff",
  "BuildRetry": false,
  "ErrorPattern": "",
  "ExcludeConsoleLog": true
}
```

### Report

|Build|Definition|Test|Pull Request|
|---|---|---|---|
|[436679](https://dev.azure.com/dnceng-public/public/_build/results?buildId=436679)|dotnet/runtime|[Microsoft.Extensions.Configuration.Binder.Tests.ConfigurationBinderTests.BindCanReadStaticProperty](https://dev.azure.com/dnceng-public/public/_build/results?buildId=436679&view=ms.vss-test-web.build-test-results-tab&runId=9707206&resultId=103476)||
|[431376](https://dev.azure.com/dnceng-public/public/_build/results?buildId=431376)|dotnet/runtime|[Microsoft.Extensions.Configuration.Binder.Tests.ConfigurationBinderTests.BindCanReadStaticProperty](https://dev.azure.com/dnceng-public/public/_build/results?buildId=431376&view=ms.vss-test-web.build-test-results-tab&runId=9539474&resultId=113305)||
|[430806](https://dev.azure.com/dnceng-public/public/_build/results?buildId=430806)|dotnet/runtime|[Microsoft.Extensions.Configuration.Binder.Tests.ConfigurationBinderTests.BindCanReadStaticProperty](https://dev.azure.com/dnceng-public/public/_build/results?buildId=430806&view=ms.vss-test-web.build-test-results-tab&runId=9525290&resultId=104530)|dotnet/runtime#93066|
|[429943](https://dev.azure.com/dnceng-public/public/_build/results?buildId=429943)|dotnet/runtime|[Microsoft.Extensions.Configuration.Binder.Tests.ConfigurationBinderTests.BindCanReadStaticProperty](https://dev.azure.com/dnceng-public/public/_build/results?buildId=429943&view=ms.vss-test-web.build-test-results-tab&runId=9503114&resultId=100624)|dotnet/runtime#92057|
|[425385](https://dev.azure.com/dnceng-public/public/_build/results?buildId=425385)|dotnet/runtime|[Microsoft.Extensions.Configuration.Binder.Tests.ConfigurationBinderTests.BindCanReadStaticProperty](https://dev.azure.com/dnceng-public/public/_build/results?buildId=425385&view=ms.vss-test-web.build-test-results-tab&runId=9360832&resultId=102709)||
|[418980](https://dev.azure.com/dnceng-public/public/_build/results?buildId=418980)|dotnet/runtime|[Microsoft.Extensions.Configuration.Binder.Tests.ConfigurationBinderTests.BindCanReadStaticProperty](https://dev.azure.com/dnceng-public/public/_build/results?buildId=418980&view=ms.vss-test-web.build-test-results-tab&runId=9192752&resultId=100603)||
|[417649](https://dev.azure.com/dnceng-public/public/_build/results?buildId=417649)|dotnet/runtime|[Microsoft.Extensions.Configuration.Binder.Tests.ConfigurationBinderTests.BindCanReadStaticProperty](https://dev.azure.com/dnceng-public/public/_build/results?buildId=417649&view=ms.vss-test-web.build-test-results-tab&runId=9151172&resultId=114738)||
#### Summary
|24-Hour Hit Count|7-Day Hit Count|1-Month Count|
|---|---|---|
|0|0|7|

 ### Known issue validation
**Build: :mag_right:** 
**Result validation: :warning:** Validation could not be done without an Azure DevOps build URL on the issue. Please add it to the "**Build: :mag_right:**" line.
**Validation performed at:** 6/28/2023 10:04:11 PM UTC

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/86532

### CODE REVIEW

**Comment by tarekgh** [comment]
@eiriktsarpalis @krwq @ericstj

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs [hunk 1]
```diff
@@ -736,37 +736,47 @@ public void GetCanReadInheritedProperties()
             Assert.Equal("Derived:Sup", options.Virtual);
         }
 
+        private static readonly object _syncLock = new object();
+
         [Fact]
         public void GetCanReadStaticProperty()
         {
-            var dic = new Dictionary<string, string>
+            // The test uses ComplexOptions.StaticProperty which is possible to be changed by other tests.
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs:744]: It is a test. We have the same class used in different scenarios.  like testing the runtime and source gen behavior with it. Even if we made it generic still need to use the exact same type in both scenarios. The issue is we need to test the static property in this class. I had other idea but didn't like it much which is mark this property as thread static which should help fix the issue, but I didn't like it as I wanted to test pure static property in the scenario.

[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs:744]: The generic parameter could just be a phantom type on the test class that you never reference inside the class itself. You could have, say:
```
class ComplexOptions<TDummy>
{
    public static int StaticProperty { get; set; }
}
```
And then instatiate in individual tests using marker types:
```C#
private class TestMarker {}

public void Test()
{
     ComplexOptions<TestMarker>.StaticProperty = 42; // Guarantees that this static property is unique
}
```

[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs:744]: Or if you prefer:
```C#
public MyComplexOptions : ComplexOptions<MyComplexOptions> { }

public void Test()
{
     MyComplexOptions .StaticProperty = 42; // Guarantees that this static property is unique
}
```

**Review by eiriktsarpalis — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/Microsoft.Extensions.Configuration.Binder/tests/Common/ConfigurationBinderTests.cs:744]: Discussed this offline.

**Pull request merged**